### PR TITLE
Feature: support for remote port forwarding

### PR DESCRIPTION
**Pull request by stankovic-marko**
This pull request implements feature request from #10.

### CODE REVIEW

**Comment by rgzr** [comment]
Hey @stankovic-marko, good job! Thank you!

I will take a look into it before merging. But at a first glance, we will be breaking existing code by adding a new parameter to `New` function. Maybe we should add a new function `NewRemote` instead of passing the `ForwardType` to keep it backwards compatible (or maybe using variadic parameter, or just a method `SetRemoteForwarding` on the `SSHTun`).

Also we should support unix sockets (this should be easily done without affecting forwarding logic).

If you feel like doing it, perfect. Otherwise I will merge your PR into an intermediate branch and work from there!

**Comment by stankovic-marko** [comment]
Hello @rgzr, thank you for your response!

I agree with everything you said, there are some things I overlooked. These things will be addressed in my next commit.

File: forward.go
Comment (rgzr) [inline forward.go]:
```suggestion
				Error: fmt.Errorf("local dial %s to %s failed: %w", tun.local.Type(), tun.local.String(), err),
```

File: forward.go
Comment (rgzr) [inline forward.go]:
```suggestion
	// Closed indicates if the connection is closed.
```

File: forward.go
Comment (rgzr) [inline forward.go]: Should be `tun.remote.Type()` for a remote port forward?

File: forward.go
Comment (rgzr) [inline forward.go:80]: Should be adjusted when it's a remote port forwarding?

File: forward.go
Comment (rgzr) [inline forward.go]: The error message should be reversed for a remote port forwarding.

File: forward.go
Comment (rgzr) [inline forward.go]: The error message should be reversed for a remote port forwarding.

File: forward.go
Comment (rgzr) [inline forward.go:119]: Is there a reason to modify this logic? I am afraid that in some corner cases one side of the connection could be closed without the other one noticing it (usually until it tries to read or write on the connection). If that happens, one of the goroutines will be blocked on the `io.Copy` call.

That was the reason why the `localConn.Close()` and `remoteConn.Close()` were just after the `connCtx` was cancelled (produced by some of the `io.Copy` fails or by stopping the SSHTun).

Also, we ensure we always close both sides of the connection (note with your changes the call to `Close` is not made if the `io.Copy` call fails) to prevent leaving open file descriptors.

Maybe I am missing something here, just talking from memory! :P Probably there's another corner case I'm missing..

File: forward.go
Comment (rgzr) [inline forward.go:125]: We should return `Closed: true` also here, in line 112.

File: sshtun.go
Comment (rgzr) [inline sshtun.go]:
```suggestion
			return tun.stop(fmt.Errorf("ssh dial %s to %s failed: %w", tun.server.Type(), tun.server.String(), err)
```

File: sshtun.go
Comment (rgzr) [inline sshtun.go]:
```suggestion
		listener, err = sshClient.Listen(tun.remote.Type(), tun.remote.String())
```

File: sshtun.go
Comment (rgzr) [inline sshtun.go]: Error message should be adapted for remote forwarding.

File: sshtun.go
Comment (rgzr) [inline sshtun.go]:
```suggestion
func (tun *SSHTun) handle(conn net.Conn) error {
```

File: sshtun.go
Comment (rgzr) [inline sshtun.go]: I think it will be better to remove this condition and add the following:

In `addConn` line 356: `if tun.forwardType == Local && tun.active == 0 {`

In `removeConn` line 375: `if tun.forwardType == Local && tun.active == 0 {`

This way the `active` field will be correctly updated.

**Review by rgzr — CHANGES_REQUESTED** [review]
I have added some minor suggestions regarding error and info messages, and a couple of ones related with logic. But in general it looks good! Thank you very much! I will add tests for this part in the future also.

File: forward.go
Comment (stankovic-marko) [reply forward.go:119]: I discovered that if multiple tunnels are chained one after the other, the order of closing connections would make one of the tunnels return an error, but I don't remember which one since I discovered this last year. :)

I remembered I explained this in a PR I submitted once: #12

File: sshtun.go
Comment (stankovic-marko) [reply sshtun.go]: Line 346 should also be updated with new variable name?

File: forward.go
Comment (stankovic-marko) [reply forward.go]: Yes, it should be reversed.

File: forward.go
Comment (stankovic-marko) [reply forward.go:125]: Yes, you are correct.

File: sshtun.go
Comment (stankovic-marko) [reply sshtun.go]: Yes, that is much better

File: forward.go
Comment (stankovic-marko) [reply forward.go]: Yes, tunneled state should be `tun.remote.Type()` for remote port forward.

**Comment by stankovic-marko** [comment]
Thank you very much for reviewing! Sorry for not polishing before submitting a PR, I implemented this a while ago and forgot what else needed to be done...

File: sshtun.go
Comment (stankovic-marko) [reply sshtun.go]: changed in 6da45df

File: forward.go
Comment (rgzr) [reply forward.go:119]: Ok, I see! I think we can investigate that problem in another PR. For now I will merge this PR in an intermediate branch, remove this part and integrate it with the main branch.
Then I will open another one to test this changes regarding connection closing to see if they don't cause other problems to other people using the library.

**Review by rgzr — APPROVED** [review]

**Comment by rgzr** [comment]
Don't worry, thank you for your contribution! I merged it to an intermediate branch removing the changes of the connection close issues and will address that problem in another PR.

**Pull request merged**
